apm-sniffer/optional-plugins/jdk-http-plugin/src/main/java/org/apache/skywalking/apm/plugin/jdk/http/define/HttpClientInstrumentation.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.jdk.http.define;
+
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.DeclaredInstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.StaticMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+import org.apache.skywalking.apm.agent.core.plugin.match.MultiClassNameMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+import static org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch.takesArgumentWithType;
+
+/**
+ * @author lican
+ */
+public class HttpClientInstrumentation extends ClassEnhancePluginDefine {
+
+    private static final String ENHANCE_HTTP_CLASS = "sun.net.www.http.HttpClient";
+
+    private static final String ENHANCE_HTTPS_CLASS = "sun.net.www.protocol.https.HttpsClient";
+
+    private static final String AFTER_METHOD = "parseHTTP";
+
+    private static final String BEFORE_METHOD = "writeRequests";
+
+    private static final String NEW_INSTANCE_METHOD = "New";
+
+    private static final String INTERCEPT_PARSE_HTTP_CLASS = "org.apache.skywalking.apm.plugin.jdk.http.HttpClientParseHttpInterceptor";
+
+    private static final String INTERCEPT_WRITE_REQUEST_CLASS = "org.apache.skywalking.apm.plugin.jdk.http.HttpClientWriteRequestInterceptor";
+
+    private static final String INTERCEPT_NEW_INSTANCE_CLASS = "org.apache.skywalking.apm.plugin.jdk.http.HttpClientNewInstanceInterceptor";
+
+    private static final String ARG_TYPE_HTTP_CONNECTION = "sun.net.www.protocol.http.HttpURLConnection";
```

Comment:
Do you really need to instrument `sun.*. HttpURLConnection`? Rather than java.net.HttpURLConnection? Or is this a protection mechenism? To avoid instrument too many unexpected classes?

Reply:
In current enhance solution, I only enhance the `sun.net.www.http.HttpClient` and `sun.net.www.protocol.https.HttpsClient` ,but I found this instrumentation will throw Exception above jdk9.so I am trying to find some new way.

Reply:
The `sun.*` classes had been renamed in JDK9+, Oracle did this.... If you want to instrument the `sun.*`, witness class should be added.

Reply:
I add a witness class that we only support jdk6,7,8 now,in jdk9+ ,the error is
```
java.lang.IllegalAccessError: superinterface check failed: class sun.net.www.http.HttpClient (in module java.base) cannot access class org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance (in unnamed module @0x555590) because module java.base does not read unnamed module @0x555590
```

Reply:
Look like a JDK9 feature. I haven't tested on jdk9 before. Is this happening in other instrumentation? Or just JDK9 bootstrap instrumentation?

Reply:
I just test bootstrap instrumentation only, I think I should test a 3rd instrumentation to verify this problem whether it exist

Reply:
Yes. I will submit a question on byte buddy.

Reply:
Submitted https://github.com/raphw/byte-buddy/issues/697

Reply:
Do you have test results for other lib instrumentation in JDK9+?

Reply:
I test okhttp3 in jdk11 ,seems works well

Reply:
> I test okhttp3 in jdk11 ,seems works well

OK. Then we only need to focus on bootstrap instrumentation, that is good. I think once the target library moved into the JDK9 as a basic requirement, this will show up.

Reply:
Please check #3194